File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    index = 1
    d = dict()
    
    for a in range(ord('A'), ord('Z') + 1):
        d[chr(a)] = index
        index += 1
    
    answer = []
    i = 0
    while i < len(msg):
        j = i + 1
        while j <= len(msg) and msg[i:j] in d:
            j += 1    
        
        if j > len(msg):
            return answer + [d[msg[i:]]]
        
        d[msg[i:j]] = index
        index += 1
        
        answer.append(d[msg[i:j-1]])
        i = j - 1
    return answer
    

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    public int[] solution(String msg) {
        Map<String, Integer> dict = new HashMap<>();
        for (int index = 1; index < 27; index++) {
            dict.put(String.valueOf((char) (index + 64)), index);
        }
    
        int left = 0;
        int next_index = 27;
        List<Integer> answer = new ArrayList<>();
        
        while (left < msg.length()) {
            int right = left + 1;
            while (right <= msg.length() && dict.containsKey(msg.substring(left, right))) {
                right++;
            }
            
            if (right == msg.length() + 1) {
                answer.add(dict.get(msg.substring(left)));
                break;
            }
            dict.put(msg.substring(left, right), next_index);
            next_index++;
            answer.add(dict.get(msg.substring(left,right-1)));
            left = right - 1;
            
        }
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
}